[WIP][Keybase] Mini-keybase implementation

Jasonyou1995 opened the pull request:
<!--
 1. Make the title of the PR is descriptive and follows this format: `[<Module>] <DESCRIPTION>`
 2. Update the _Assigness_, _Labels_, _Projects_, _Milestone_ before submitting the PR for review.
 3. Add label(s) for the purpose (e.g. `persistence`) and, if applicable, priority (e.g. `low`) labels as well.
-->

## Description

Implementing Mini ED-25519 Keybase:

 Code complete implementation of:
 - [x] CREATE and store keys
 - [x] DELETE keys
 - [ ] Document the key architecture

## Issue

Fixes #150 

## Type of change

Please mark the relevant option(s):

- [x]  New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [x] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

<!-- List out all the changes made-->

- Implementing the mini keybase package
- Create Key
- Store Key
- Delete Key

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`
<!--
If you added additional tests or infrastructure, describe it here.

Bonus points for images and videos or gifs.
-->

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

deblasis committed:
refactor(Shared): RPC config defaults -changing soon,I'm centralizin

deblasis committed:
fix(Utility): RPC: updated to use test_artifacts defaults

deblasis committed:
feat(Utility): CLI: updated to use test_artifacts default

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
docs(Utility): RPC and node basic docs

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

deblasis committed:
docs(Utility): CLI: Changelog

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

deblasis committed:
fix(Utility): CLI: fixed output to user. It shouldn't be logging

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

deblasis committed:
fix(Utility): CLI code review feedback

deblasis committed:
style(Utility): code review feedback

deblasis committed:
feat(Tooling): Updated makefile commands to better handle codegen

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
feat(Tooling): Updated makefile commands to better handle codegen

deblasis committed:
fix(Utility): Fix duplicated models

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
chore(Utility): fixed versioning schema

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

deblasis committed:
feat(Utility): CLI documentation generator + first stab at docs

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

deblasis committed:
fix(Utility): RPC configs post-merge

deblasis committed:
fix(Utility): RPC configs post-merge

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/utility_local_proof_of_stake_cli_CLI

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
fix(Utility): CLI specific fixes

deblasis committed:
fix(Utility): types fixes

deblasis committed:
feat(Consensus): configOptions

deblasis committed:
feat(Consensus): configOptions

deblasis committed:
feat(P2P): configOptions

deblasis committed:
feat(P2P): configOptions

deblasis committed:
feat(Utility): CLI: using configOptions to inject PK

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/utility_local_proof_of_stake_cli_RPC

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
fix(Utility): RPC fix post merge

deblasis committed:
fix(Utility): RPC disabled by default because of TECHDEBT

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

deblasis committed:
chore(go.mod): tidy

deblasis committed:
Merge branch 'main' into issue/utility_local_proof_of_stake_cli_CLI

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/utility_local_proof_of_stake_cli_CLI

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
fix(go.mod): tidy

deblasis committed:
fix(CLI): test_artifacts

deblasis committed:
fix(go.mod): tidy

deblasis committed:
fix(CLI): updated to use new typesUtil.ActorType

deblasis committed:
fix(CLI): runtime based init

deblasis committed:
fix(test_artifacts): fix

deblasis committed:
fix(CLI): runtime using WithRandomPK()

deblasis committed:
fix(CLI): fixed client-only initialization

deblasis committed:
fix(Makefile): protogen first

deblasis committed:
feat(Utility): GetActorName function (can we autogenerate these?)

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/utility_local_proof_of_stake_cli_RPC

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
fix(RPC): test_artifacts in runtime

deblasis committed:
fix(go.mod): tidy

deblasis committed:
fix(CLI): debug commands are now feature flagged

deblasis committed:
chore(go.mod): tidy

deblasis committed:
chore(CLI): git rm app/pocket/rpc/client.gen.go

deblasis committed:
chore(CLI): s/j/tx and s/prepareTx/prepareTxJson

deblasis committed:
refactor(shared): converters

deblasis committed:
fix(CLI): debug nits

deblasis committed:
refactor(CLI): confirmation and credentials

deblasis committed:
chore(CLI): removed HACK todo

deblasis committed:
fix(Makefile): premature protoc

deblasis committed:
fix(RPC): added imports used in codegen files

deblasis committed:
fix(RPC): added imports used in codegen files

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
docs(RPC): added swagger editor link

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
feat(RPC): config proto

deblasis committed:
feat(RPC): RPCModule and noopRpcModule

deblasis committed:
refactor(Shared): shared.Create -> shared.CreateNode

deblasis committed:
refactor(RPC): moved scaffolding into rpc module

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
fix(RPC): removed redundant file

deblasis committed:
fix(CLI): debug merge fix

deblasis committed:
docs(RPC): fixed link after refactoring

deblasis committed:
docs(RPC): updated code organization post refactoring

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

deblasis committed:
Update app/client/cli/utils.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
fix(RPC): gitignoring generated files

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
fix(RPC): gitignoring generated files

deblasis committed:
style(Persistence): reverting space change

deblasis committed:
refactor(Defaults): runtime/defaults package

deblasis committed:
chore(CLI): issue handle

deblasis committed:
chore(CLI): Issue #310 links

deblasis committed:
refactor(CLI): preallocation fix

deblasis committed:
style(CLI): oneMillion

deblasis committed:
style(CLI): s/RelayChainIDs/relayChainIDs

deblasis committed:
feat(Utility): ActorType.GetName()

deblasis committed:
chore(Utility): tracking issue #142

deblasis committed:
chore(Utility): GetBytes -> GetCanonicalBytes

deblasis committed:
chore(CLI): actorCmd -> accountCmd

deblasis committed:
docs(CLI): fromAddr note (address currently sourced from pk)

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/utility_local_proof_of_stake_cli_CLI

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
chore(Runtime): new default value from main

deblasis committed:
refactor(CLI): moving utility functions in utils.go

deblasis committed:
chore(CLI): NewDebug -> NewDebugCommand

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/utility_local_proof_of_stake_cli_RPC

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
refactor(Consensus): mocks with Return(nil) and not Do(...)

deblasis committed:
refactor(Consensus): mocks with Return(nil) and not Do(...)

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
docs(RPC): updated changelog versions

deblasis committed:
fix(Makefile): generate_rpc_openapi

deblasis committed:
fix(Makefile): fix for git clone + make develop_test

deblasis committed:
Update rpc/v1/openapi.yaml

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' of github.com:deblasis/pocket into issue/utility_local_proof_of_stake_cli_RPC

deblasis committed:
Update rpc/doc/CHANGELOG.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Update rpc/noop_module.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' of github.com:deblasis/pocket into issue/utility_local_proof_of_stake_cli_RPC

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
chore(Shared): added TODOes for ValidateXXX() in modules

deblasis committed:
docs(RPC): broadcast_tx_sync summary updated

deblasis committed:
Update rpc/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Update rpc/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Update rpc/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' of github.com:deblasis/pocket into issue/utility_local_proof_of_stake_cli_RPC

deblasis committed:
fix(Makefile): gitignoring generated files breaks tests fix

deblasis committed:
docs(CLI): added todos for exactArgs

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/utility_local_proof_of_stake_cli

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
chore(RPC): added types.go

deblasis committed:
refactor(RPC): sourcing defaults from defaults not test_artifacts

deblasis committed:
feat(CLI): system commands RPC🤝CLI

deblasis committed:
feat(Consensus): CurrentRound() and CurrentStep()

deblasis committed:
feat(RPC): /v1/consensus/round_state

deblasis committed:
feat(RPC): /v1/consensus/round_state

deblasis committed:
feat(RPC): /v1/consensus/round_state handler

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

deblasis committed:
refactor(CLI): system command typoes copypastas

deblasis committed:
feat(CLI): consensus commands -RoundState and individual ones

deblasis committed:
chore(CLI): typo

deblasis committed:
docs(CLI): short commands descriptions

deblasis committed:
docs(Config): Adding some more color around configuration

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

deblasis committed:
fix(CLI): tx message signing

deblasis committed:
fix(CLI): tx message signing

deblasis committed:
feat(CLI): reporting statuscode and body

deblasis committed:
feat(CLI): reporting statuscode and body

deblasis committed:
fix(Proto): deterministic

deblasis committed:
refactor(CLI): prepareTxJSON -> prepareTxBytes

deblasis committed:
docs(Docs): Adding some more color config + raw_hex_bytes

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

deblasis committed:
refactor(CLI): Stake command

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

deblasis committed:
feat(Tooling): swagger-ui

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
feat(Tooling): swagger-ui

deblasis committed:
feat(RPC): cors feature flag

deblasis committed:
Update utility/types/message.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
chore(Runtime): comment spacing

deblasis committed:
chore(Runtime): comment spacing

deblasis committed:
docs(CLI): Changelog

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
docs(CLI): changelog

deblasis committed:
docs(RPC): changelog

deblasis committed:
docs(RPC): changelog

deblasis committed:
Update runtime/docs/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
chore(Consensus): SetBus mock Do(func(modules.Bus) {}) -> Return()

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
chore(Consensus): SetBus mock Do(func(modules.Bus) {}) -> Return()

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' of github.com:deblasis/pocket into issue/utility_local_proof_of_stake_cli_RPC

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_CLI' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

deblasis committed:
docs(Pocket): changelog

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

deblasis committed:
Update app/pocket/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Update app/pocket/doc/README.md

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' of github.com:deblasis/pocket into issue/utility_local_proof_of_stake_cli_RPC

deblasis committed:
docs(RPC): docs TOC

deblasis committed:
docs(RPC): Transports -> Endpoints

deblasis committed:
docs(RPC): swagger ui link to editor and ref to make cmd

deblasis committed:
feat(rpcServer): rpcServer is now IntegratableModule

deblasis committed:
chore(Shared): tracking TODO (implement validations) #334

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
Merge remote-tracking branch 'origin/main' into issue/utility_local_proof_of_stake_cli_RPC

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
fix(RPC): merge fix

deblasis committed:
chore(go.mod): tidy

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
chore(go.mod): tidy

deblasis committed:
feat(Tooling): added empty mock_module package for cold start

deblasis committed:
Merge remote-tracking branch 'upstream/main' into issue/utility_local_proof_of_stake_cli_RPC

Signed-off-by: Alessandro De Blasis <alex@deblasis.net>

deblasis committed:
docs(RPC): nit

deblasis committed:
Merge branch 'issue/utility_local_proof_of_stake_cli_RPC' into issue/utility_local_proof_of_stake_cli

Jasonyou1995 committed:
Adding the keybase package

Jasonyou1995 committed:
Adding relevant building blocks

Updates and modifications needed

Olshansk commented:
@Jasonyou1995 I see that you added myself and Andrew as reviewers. Are you looking for us to review this PR?

deblasis committed:
Update app/client/cli/consensus.go

Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

deblasis committed:
docs(CLI): better commands descriptions

Jasonyou1995 committed:
Adding badger DB for keybase

Jasonyou1995 committed:
Package name update

deblasis committed:
feat(CLI): boldText helper

deblasis committed:
style(CLI): nit: real estate

Jasonyou1995 commented:
> 3\. Th

Thanks for the suggestions @Olshansk ! I will checkout the branches created by @deblasis locally. 

The codes I'm working on are not yet ready for review. I will switch from draft to formal PR once the codes are ready. I will specify the parts that need to be reviewed.  Thanks for the good examples!

deblasis committed:
refactor(CLI): ConsensusState

deblasis committed:
docs(CLI): updated docs

Jasonyou1995 committed:
rename deprecated codes to keybase_test

Integrating to working keybase package later

Jasonyou1995 committed:
Adding keybase CLI package initial framework

Jasonyou1995 committed:
update `create` and `delete` commands descriptions

Jasonyou1995 committed:
Changing package name from "keybase" to "keys"

Avoiding naming conflict with keybase.app

Jasonyou1995 committed:
Update delete command comments

Jasonyou1995 committed:
Update root command comments

Olshansk commented:
@Jasonyou1995 If you change the base, it'll help us only see the differences. Once the other PR is merged, we can change it back.

![Screen Shot 2022-11-09 at 11 10 26 AM](https://user-images.githubusercontent.com/1892194/200919945-ed55b4bd-8d16-4310-9bc3-e770ed7ecaae.png)

Also, an idea I had is adding a target to our `Makefile` that can be invoked with `make create_key` that'll be an easy entrypoint for the CLI call with some defaults in place. Not a must, but it'd be an easy way for us to pull this branch down and test out the code locally.

Jasonyou1995 committed:
adding mnemonic sub-command to keys

Jasonyou1995 commented:
> @Jasonyou1995 If you change the base, it'll help us only see the differences. Once the other PR is merged, we can change it back.
> 
> ![Screen Shot 2022-11-09 at 11 10 26 AM](https://user-images.githubusercontent.com/1892194/200919945-ed55b4bd-8d16-4310-9bc3-e770ed7ecaae.png)
> 
> Also, an idea I had is adding a target to our `Makefile` that can be invoked with `make create_key` that'll be an easy entrypoint for the CLI call with some defaults in place. Not a must, but it'd be an easy way for us to pull this branch down and test out the code locally.

Thanks for the commend @Olshansk! I just changed to the branch to `issue/utility_local_proof_of_stake_cli_RPC`. But I rebased to `deblasis:issue/utility_local_proof_of_stake_cli` before so there are still some delta. Do you want me to rebase it back to the "pocket/" branch?

For the `Makefile` do you want the new target be in a different PR?

Jasonyou1995 committed:
updating add key flags

Olshansk commented:
> For the Makefile do you want the new target be in a different PR?

Same PR since its related. It's the tool we can use to verify your chang.

> Thanks for the commend @Olshansk! I just changed to the branch to issue/utility_local_proof_of_stake_cli_RPC. But I rebased to deblasis:issue/utility_local_proof_of_stake_cli before so there are still some delta. Do you want me to rebase it back to the "pocket/" branch?

The goal is really just to make sure that it's easy for the reviewer to see the changes, regardless of how we get there. In this specific case. I believe that @deblasis will merge #169 soon, so:

1. That gets merged in
2. Merge this with maser
3. Resolve conflicts
4. Push up changes
5. Make the base of this PR main
6. Let us, the reviewers, know the status of this PR or where you might need preliminary feedback (sync or async)

Jasonyou1995 committed:
create functions update

Jasonyou1995 committed:
Resolve cmd syntax issues

Jasonyou1995 committed:
Replace deprecated dependencies from Cosmos-SDK

replace github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1

Jasonyou1995 committed:
Update mnemonic flag comments

Jasonyou1995 committed:
fix issues in keys create package

Jasonyou1995 committed:
Update go.mod

Jasonyou1995 committed:
Update delete.go

Jasonyou1995 committed:
Update .gitignore to ignore `.keybase files`

Jasonyou1995 committed:
Adding level.db support for keybase

Jasonyou1995 committed:
level.db close

Jasonyou1995 committed:
implement delete from local keybase

Jasonyou1995 committed:
Adding create_key to Makefile

Jasonyou1995 committed:
Clean comments

Jasonyou1995 commented:
> Any chance of minimizing this PR? I had expected the package and integration to be separate issues

Thank you for clarifying @andrewnguyen22 ! I think that's a good idea to separate so we can have a clear view to the changes.

I would like to get some of your feedbacks based on the current functions for Create, Delete, and Mnemonic. I will make some change accordingly.

Jasonyou1995 commented:
### Makefile `create_key` test | Delete key output
<img width="1904" alt="Screen Shot 2022-11-10 at 5 40 43 PM" src="https://user-images.githubusercontent.com/15693947/201221452-aa37f5b8-2121-4bb4-81a2-b2ca2a3c3bb1.png">

### Delete none-exist key
<img width="1904" alt="Screen Shot 2022-11-10 at 5 41 02 PM" src="https://user-images.githubusercontent.com/15693947/201221456-d0b94735-0d86-4975-9e9e-e537ebf2555a.png">

### Create a new key with given name
<img width="1904" alt="Screen Shot 2022-11-10 at 5 41 28 PM" src="https://user-images.githubusercontent.com/15693947/201221458-b6fae2af-a974-4456-a723-6f570cffa1eb.png">


Please let me know if there is change needed for the output format and features. There are some TODO listed in the source so we can discuss which ones need to be updated.

**PS: my updates are all in `share/keys/cmd**

Jasonyou1995 committed:
BIP39 Mnemonic generating in create command

Jasonyou1995 committed:
Adding TODO for generating key with PocketCrypto

Jasonyou1995 committed:
Add key type for keybase in create.go

Jasonyou1995 committed:
Using shared/crypto/ed25519.go to generate keys

andrewnguyen22 commented on shared/keys/cmd/create.go:
```diff
@@ -0,0 +1,163 @@
+
+	// TODO: Using Mnemonic to generate keys
+
+	// TODO: Import the CryptoPocket!
```
`TODO: (<Issue-#>): Import CryptoPocket`

Please add follow up issue numbers - also, why can't this be done `IN_THIS_COMMIT`?

Jasonyou1995 replied:
I just have the crypto functions updated, and this is done in this commit now.

andrewnguyen22 commented on shared/keys/cmd/delete.go:
```diff
@@ -0,0 +1,94 @@
+	rootCmd.AddCommand(deleteCmd)
+
+	// Local flags
+	//f := deleteCmd.Flags()
```
Remove comments or add comment for future use

Jasonyou1995 replied:
Comments removed

andrewnguyen22 commented on shared/keys/cmd/root.go:
```diff
@@ -0,0 +1,95 @@
+func init() {
+	cobra.OnInitialize(initConfig)
+
+	// Here you will define your flags and configuration settings.
```
Please remove example comments and provide source

Jasonyou1995 committed:
Fixing nil pointer issue during recovery flag mode

Issue: using invalid reference to cosmos context

Jasonyou1995 committed:
Fixing private key generate error

Jasonyou1995 committed:
Update print functions and command instructions

Jasonyou1995 commented on shared/keys/cmd/create.go:
```diff
@@ -0,0 +1,163 @@
+/*
```
Thanks pointing out! I was not so sure about this neither. That license and copyright sections were generated by Cobra automatically.

Jasonyou1995 commented on shared/keys/LICENSE:
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Yes, that is correct! I rebased to @deblasis 's latest director, and that's why it shows all the changes.

Jasonyou1995 commented on shared/keys/cmd/create.go:
```diff
@@ -0,0 +1,163 @@
+	FlagPublicKey = "pubkey"
+
+	//// For output formats
+	//OutputFormatText = "text"
```
Just updated and removed the unnecessary comments with future TODO

Jasonyou1995 commented on shared/keys/cmd/create.go:
```diff
@@ -0,0 +1,163 @@
+	name := args[0]
+
+	// TODO: determine proper keystore location later
+	kb, err := leveldb.OpenFile("./.keybase/poktKeys.db", nil)
```
Added error handling.

Jasonyou1995 commented on shared/keys/cmd/create.go:
```diff
@@ -0,0 +1,163 @@
+	kb, err := leveldb.OpenFile("./.keybase/poktKeys.db", nil)
+
+	// Creating key with ED25519
+	pubKey, privKey, err := ed25519.GenerateKey(nil) // TODO: using mnemonic to generate (read)
```
Error handled

Jasonyou1995 commented on shared/keys/cmd/create.go:
```diff
@@ -0,0 +1,163 @@
+	var mnemonic string
+	var bip39Passphrase string = ""
+
+	recover, _ := cmd.Flags().GetBool(flagRecover)
```
Error handled

Jasonyou1995 committed:
Removing comments and update changes

Jasonyou1995 committed:
Updating reviewed changes

Jasonyou1995 committed:
Removing automatically generated MIT licenses

Jasonyou1995 commented:
Just update codes based on review from @andrewnguyen22 . 

Currently, the module is fully isolated from the `pocket` module, so there is still some build error. Could you suggest how to update the module/package structure change?

I copied the `shared/crypto/` folder to my local directory, but I think that's redundant, so if you could suggest a way to update that would be great!

I used the functions `ED25519.go` to generate keys, so please check the validity of the key generating process in case anything is wrong.

Jasonyou1995 committed:
Delete LICENSE

Jasonyou1995 committed:
Update create.go

Jasonyou1995 committed:
Update create.go

Olshansk commented:
> Currently, the module is fully isolated from the pocket module, so there is still some build error. Could you suggest how to update the module/package structure change?

Could you provide more details (screenshot? error) on what the build error is so we can point you in the right direction?

> I used the functions ED25519.go to generate keys, so please check the validity of the key generating process in case anything is wrong.

What do you mean by "validity" in this context?

> The mnemonic phrases is currently stored in the level.db along with other key information in JSON format. Please let me know if you want to remove any info from the key storage.

I'd personally store mnemonic phrases  in some dot file on the host machine, but I haven't looked into standard patterns. Does cosmos or do other projects use leveldb for it?

Olshansk commented on shared/keys/main.go:
```diff
@@ -0,0 +1,9 @@
+package main
```
@Jasonyou1995 When we started working on #320, we ran into an issue where all methods in a directory and its subdirectories were different packages, because every directory in a Go package. Ref: https://www.codementor.io/@inanc/understanding-go-packages-cgsx1hzrx

This is a separate project (not a separate package), which needs a new main file, go.sum, go.mod, and there should only be one in any project. For example, VRF (https://github.com/pokt-network/pocket/tree/main/consensus/leader_election/vrf) is a package in the same project that is not integrated into the end-to-end flow.

Lmk if that makes sense. I can review the core work once this change is made.

Jasonyou1995 replied:
Yes! I will update the code for a better re-usability and merge all code from key-base into the pocket project. Thanks for pointing out and sharing the article!

Jasonyou1995 commented on Makefile:
```diff
@@ -145,6 +146,12 @@ build_and_watch:
 compose_and_watch: docker_check db_start monitoring_start
 	docker-compose -f build/deployments/docker-compose.yaml up --force-recreate node1.consensus node2.consensus node3.consensus node4.consensus
 
+# TODO (jasonyou): create key and output proper data format and entry point
```
I added it for the next PR so we don't forget.

Jasonyou1995 commented on .gitignore:
```diff
@@ -29,7 +29,8 @@ prototype/vendor/
 # Generated Mocks (for testing)
-shared/modules/mocks/
+shared/modules/mocks/*
+!shared/modules/mocks/mocks.go
```
I think that is changed by @deblasis

Jasonyou1995 commented on Makefile:
```diff
@@ -145,6 +146,12 @@ build_and_watch:
+.PHONY: create_key
+## build and install the shared/keys module and invoke key creation
+create_key:
+	cd ./shared/keys; go install; cd ../..; keys create test_key
```
Yeah you are right, I'm updating the package structure so it can have all the dependencies live in a single go.mod file

Jasonyou1995 reviewed (commented):
Reviewed items been updated or is updating

Jasonyou1995 committed:
Move keys to root directory as a separate project

andrewnguyen22 commented on keys/cmd/delete.go:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	// Check if the key name is in the DB
+	_, err = kb.Get([]byte(name), nil)
```
Non-Blocker; Nit

Make inline (see example above)

andrewnguyen22 commented on keys/cmd/delete.go:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	// remove key based on name KEY ID
+	err = kb.Delete([]byte(name), nil)
```
Non-Blocker; Nit

Make inline (see example above)

andrewnguyen22 commented on keys/cmd/delete.go:
```diff
@@ -0,0 +1,72 @@
+
+	fmt.Printf("Key (%s) deleted!\n", name)
+
+	defer kb.Close()
```
`defer` means 'execute at the conclusion of the function

Thus, `defer kb.Close()` should be at line 40

andrewnguyen22 commented on keys/cmd/mnemonic.go:
```diff
@@ -0,0 +1,89 @@
+		var entropySeed []byte
+
+		// get entropy from users if --unsafe-entropy flag is passed
+		if userEntropy, _ := cmd.Flags().GetBool(flagUserEntropy); userEntropy {
```
Handle error

andrewnguyen22 commented on keys/cmd/delete.go:
```diff
@@ -0,0 +1,72 @@
+	Long: `Delete the public key from the backend keystore offline
+Note: Delete key does not delete private key stored in a ledger device.
+`,
+	Args: cobra.MinimumNArgs(1),
```
Try `ExactArgs` instead of `MinimumArgs`

andrewnguyen22 commented on keys/cmd/create.go:
```diff
@@ -0,0 +1,163 @@
+		return err
+	}
+
+	defer kb.Close()
```
Ditto issue with `defer` (see below) watch out for this in other usages

andrewnguyen22 commented on keys/cmd/create.go:
```diff
@@ -0,0 +1,163 @@
+// - confirmation from user for overriding existing key
+// - implement key phrase intput from user secure keys
+func runAddCmd(cmd *cobra.Command, args []string) error {
+	var err error
```
Remove unnecessary `err` declaration

andrewnguyen22 commented on keys/cmd/create.go:
```diff
@@ -0,0 +1,163 @@
+)
+
+// createCmd represents the create command
+var createCmd = &cobra.Command{
```
I'm not convinced the `CLI` portion of the keybase should be in this pacakge and not in the `app/client/cli` folder. WYT @Olshansk @deblasis

andrewnguyen22 commented on keys/cmd/create.go:
```diff
@@ -0,0 +1,163 @@
+	address := privateKey.Address()
+
+	// JSON encoding
+	var keystore = key{name, publicKey, privateKey, address, mnemonic}
```
It appears as if we are storing `clear text` `non obfuscated` private keys in a leveldb. This is not up to the V0 implementation's security requirements. 

[V0 Keybase Implementation](https://github.com/pokt-network/pocket-core/blob/40c52a1cf3d4ccc6ed161b2ac1853a003c32324f/crypto/keys/keybase.go#L238)

andrewnguyen22 reviewed (changes requested):
A few nits, a few blockers. The biggest concern for me is the lack of security parity with V0. See comments below

Jasonyou1995 committed:
remove keys binary

Jasonyou1995 committed:
Update .gitignore

Jasonyou1995 committed:
Update go.mod

Jasonyou1995 committed:
Update go.sum

Jasonyou1995 committed:
Update reviewed changes

* Incline Error update
* defer location update for DB close

Jasonyou1995 committed:
Refactoring /keys to /app/client/cli/keys

Jasonyou1995 reviewed (commented):
Updated comments from @Olshansk and @andrewnguyen22 . One blocker.

deblasis commented:
@Jasonyou1995 regarding the `go.mod` and dependencies issues I have pulled your branch, made a few changes and now your logic is connected to the CLI. 

You can scan the delta of the changes here: 
https://github.com/pokt-network/pocket/compare/150-keybase-miniature-keybase-implementation-to-enable-m1-create-and-delete...150-keybase-miniature-keybase-implementation-to-enable-m1-create-and-delete_ADB_POC

**Do not merge it to your branch** 🙂 use it as an inspiration to understand what didn't work before and why. 

Bear in mind that I am simply addressing the cobra part and simply "made it work" in the fastest way possible to unlock you at least on the dependencies front, everything else still needs to be addressed. 

PTAL and revert with any questions you might have.

Jasonyou1995 committed:
Implementing core security functions

encrypt/decrypt with AES
generateKeyFromPassPhrase
generateRandomKey

Jasonyou1995 committed:
Remove redundant keys package

Jasonyou1995 committed:
Update the keys CLI cobra structure according reviews from @deblasis

Jasonyou1995 committed:
Update Makefile

Jasonyou1995 committed:
Merge branch 'main' into 150-keybase-miniature-keybase-implementation-to-enable-m1-create-and-delete

Jasonyou1995 committed:
Update go.mod

Ran `go mod tidy`

Jasonyou1995 committed:
Remove redundant crypto package from keys

Jasonyou1995 committed:
Update keys.go

Jasonyou1995 committed:
Save utility functions to a different script

Jasonyou1995 committed:
Remove mnemonic cmd

Jasonyou1995 committed:
Secure private key with AES encryption and define secure local keybase path

* Generating 32 bytes keys from user input to init AES cipher
* Update keybase location
* Output format update: ignore mnemonic and encrypted private key

Jasonyou1995 committed:
Update user confirmation when key exists

* Concise error check
* Keybase connection update

Jasonyou1995 committed:
Implement --recover and -i interactive flags

Jasonyou1995 committed:
Update .gitignore

Jasonyou1995 committed:
Update Delete DB setup

Jasonyou1995 committed:
Update CHANGELOGS and documents

Olshansk commented:
@Gustavobelfort No need to look at this right now, but just wanted to provide context that this work exists as a reference or starting point.
